src/tracers.jl

```diff
@@ -12,147 +12,142 @@ The provided index set type `S` has to satisfy the following conditions:
 Subtypes of `AbstractSet{<:Integer}` are a natural choice, like `BitSet` or `Set{UInt64}`.
 """
 
+empty_sparse_vector(T) = T()
+empty_sparse_matrix(T) = T()
+
+sparse_vector(T, index) = T([index])
+sparse_vector(::Type{T}, index) where {T<:DuplicateVector} = T(index)
+sparse_vector(::Type{T}, index) where {T<:SortedVector} = T(index)
+
+#================#
+# Elementwise OR #
+#================#
+
+## We use ∨ to represent the elementwise OR
+# TODO REVIEW: \vee looks too much like the character 'v' in code
+# Gradient representations
+∨(a::G, b::G) where {G<:AbstractSet} = a ∪ b
+∨(a::G, b::G) where {G<:DuplicateVector} = G(vcat(a.data, b.data))
+∨(a::G, b::G) where {G<:SortedVector} = a ∪ b
+∨(a::G, b::G) where {G<:RecursiveSet} = a ∪ b
```

Comment:
This can go in other files. Basically the central `tracker.jl` file should read as if it had been written for `AbstractVector{Bool}` or `AbstractMatrix{Bool}`, give or take

Reply:
In which case:
```julia
elementwise_or(a::AbstractVector{Bool}, b::AbstractVector{Bool}) = a .| b
outer_product_or(a::AbstractVector{Bool}, b::AbstractVector{Bool}) = a .| transpose(b)
```

Reply:
Here, the naked theory happens. Save the magic for the set implementations

Reply:
Ok, we need to make a call here. 
The cleanest abstraction would be to make everything either a `AbstractSet{<:Integer}` or an `AbstractVector{<:Bool}`.

Based on this decision, we get either `a ∪ b` or  `a .| b` everywhere. Either one single method
```Julia
∨(a::G, b::G) where {G<:AbstractSet{<:Integer}} = a ∪ b
```
or one single method
```Julia
∨(a::G, b::G) where {G<:AbstractVector{Bool}} = a .| b
```
So far, we have only been using index-sets to represent the *mathematical* concept of a sparse vector. I therefore think `AbstractSet{<:Integer}` is the better abstraction. 

Otherwise, we would have to introduce a wrapper type that turns the index-sets we use into `AbstractVector{Bool}`. And element-wise `|` might be very inefficient on an `AbstractVector` which is actually an `AbstractSet` in disguise.

Reply:
see above, you're right, let's go `AbstractSet{<:Integer}` all the way (even if we don't implement all the features of an abstract set

Reply:
and `AbstractSet{Tuple{I,I}}` for Hessians?